File: src-tauri/src/parsing/entry_parser.rs
/// Parse a raw JSONL entry into a `ParsedMessage`.
///
/// Content-shape and usage normalization live in `content_normalization`;
/// tool-call/result extraction lives in `tool_extraction`. This file owns
/// the entry-level dispatcher that decides which message kind we're
/// looking at and how to wire those helpers together.
use crate::types::jsonl::RawJsonlEntry;
use crate::types::messages::{ParsedMessage, ParsedMessageContent, SystemEventData};

use super::content_normalization::{parse_message_content, parse_usage};
use super::system_event::build_system_event_data;
use super::tool_extraction::{extract_tool_calls, extract_tool_results};

const KNOWN_TYPES: &[&str] = &[
    "user",
    "assistant",
    "system",
    "summary",
    "file-history-snapshot",
    "queue-operation",
    "progress",
];

/// Returns None for entries without uuid or unknown types.
pub fn parse_entry(entry: &RawJsonlEntry) -> Option<ParsedMessage> {
    if !KNOWN_TYPES.contains(&entry.entry_type.as_str()) {
        return None;
    }

    let uuid = match entry.uuid.as_ref() {
        Some(u) if !u.is_empty() => u.clone(),
        _ => {
            if entry.entry_type == "progress" {
                let tool_id = entry.tool_use_id_ref.as_deref().unwrap_or("unknown");
                let ts = entry.timestamp.as_deref().unwrap_or("0");
                format!("progress-{}-{}", tool_id, ts)
            } else {
                return None;
            }
        }
    };

    let is_conversational = matches!(entry.entry_type.as_str(), "user" | "assistant" | "system");

    let mut content = ParsedMessageContent::Text(String::new());
    let mut role = None;
    let mut usage = None;
    let mut model = None;
    let mut request_id = None;
    let mut is_meta = entry.is_meta.unwrap_or(false);
    let mut is_compact_summary = entry.is_compact_summary.unwrap_or(false);
    let mut subtype: Option<String> = None;
    let mut event_data: Option<SystemEventData> = None;

    if is_conversational {
        if let Some(ref msg_value) = entry.message {
            match entry.entry_type.as_str() {
                "user" => {
                    if let Some(msg_content) = msg_value.get("content") {
                        content = parse_message_content(msg_content);
                    }
                    role = msg_value
                        .get("role")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());
                }
                "assistant" => {
                    if let Some(msg_content) = msg_value.get("content") {
                        content = parse_message_content(msg_content);
                    }
                    role = msg_value
                        .get("role")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());
                    model = msg_value
                        .get("model")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());

                    if let Some(usage_val) = msg_value.get("usage") {
                        usage = Some(parse_usage(usage_val));
                    }

                    request_id = entry.request_id.clone();
                }
                "system" => {
                    is_meta = entry.is_meta.unwrap_or(false);
                    if let Some(ref sub) = entry.subtype {
                        subtype = Some(sub.clone());
                        event_data = build_system_event_data(entry);
                    }
                }
                _ => {}
            }
        }

        if entry.entry_type == "user" {
            if let Some(true) = entry.is_compact_summary {
                is_compact_summary = true;
            }
        }
    }

    if entry.entry_type == "progress" {
        subtype = Some("progress".to_string());
        if let Some(ref data) = entry.data {
            if let Some(msg) = data.get("message").and_then(|v| v.as_str()) {
                content = ParsedMessageContent::Text(msg.to_string());
            }
        }
    }

    if entry.entry_type == "queue-operation" {
        subtype = Some("queue_operation".to_string());
        event_data = Some(SystemEventData {
            subtype: "queue_operation".to_string(),
            operation: entry.operation.clone(),
            queued_content: entry.content.clone(),
            ..Default::default()
        });
    }

    let tool_calls = extract_tool_calls(&content);
    let tool_results = extract_tool_results(&content);

    // Synthetic fallback when an entry lacks a timestamp. Match Go's
    // `time.Now().UTC().Format(time.RFC3339)` (entry_parser.go:111): seconds
    // precision + trailing `Z`. Plain `to_rfc3339()` emits nanos + `+00:00`,
    // which diverges. This field derives from `now()`, so it is never
    // golden-diffed — only the format shape is asserted (see tests).
    let timestamp = entry.timestamp.clone().unwrap_or_else(|| {
        chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true)
    });

    Some(ParsedMessage {
        uuid: uuid.clone(),
        parent_uuid: if is_conversational {
            entry.parent_uuid.clone()
        } else {
            None
        },
        message_type: entry.entry_type.clone(),
        timestamp,
        role,
        content,
        usage,
        model,
        cwd: entry.cwd.clone(),
        git_branch: entry.git_branch.clone(),
        agent_id: entry.agent_id.clone(),
        is_sidechain: entry.is_sidechain,
        is_meta,
        user_type: entry.user_type.clone(),
        tool_calls,
        tool_results,
        source_tool_use_id: entry.source_tool_use_id.clone(),
        source_tool_assistant_uuid: entry.source_tool_assistant_uuid.clone(),
        tool_use_result: entry.tool_use_result.clone(),
        is_compact_summary: if is_compact_summary { Some(true) } else { None },
        request_id,
        subtype,
        event_data,
    })
}

#[cfg(test)]
#[path = "entry_parser_tests.rs"]
mod entry_tests;

#[cfg(test)]
#[path = "entry_parser_event_tests.rs"]
mod entry_event_tests;

File: internal/parsing/entry_parser.go
package parsing

import (
	"encoding/json"
	"slices"
	"time"

	"claude-devtools/internal/domain"
)

var knownTypes = []string{
	"user", "assistant", "system", "summary",
	"file-history-snapshot", "queue-operation", "progress",
}

// parseEntry mirrors entry_parser::parse_entry. Returns nil for entries with no
// uuid (except progress, which synthesizes one) or unknown types.
func parseEntry(e *domain.RawJsonlEntry) *domain.ParsedMessage {
	if !slices.Contains(knownTypes, e.EntryType) {
		return nil
	}

	var uuid string
	if e.UUID != nil && *e.UUID != "" {
		uuid = *e.UUID
	} else if e.EntryType == "progress" {
		toolID := "unknown"
		if e.ToolUseIDRef != nil {
			toolID = *e.ToolUseIDRef
		}
		ts := "0"
		if e.Timestamp != nil {
			ts = *e.Timestamp
		}
		uuid = "progress-" + toolID + "-" + ts
	} else {
		return nil
	}

	isConversational := e.EntryType == "user" || e.EntryType == "assistant" || e.EntryType == "system"

	content := textContent("")
	var role, model, requestID, subtype *string
	var usage *domain.TokenUsage
	var eventData *domain.SystemEventData
	isMeta := boolOr(e.IsMeta)
	isCompactSummary := boolOr(e.IsCompactSummary)

	if isConversational && len(e.Message) > 0 {
		var mv map[string]json.RawMessage
		_ = json.Unmarshal(e.Message, &mv)
		switch e.EntryType {
		case "user":
			if c, ok := mv["content"]; ok {
				content = parseMessageContent(c)
			}
			if r, ok := decodeString(mv["role"]); ok {
				role = &r
			}
		case "assistant":
			if c, ok := mv["content"]; ok {
				content = parseMessageContent(c)
			}
			if r, ok := decodeString(mv["role"]); ok {
				role = &r
			}
			if m, ok := decodeString(mv["model"]); ok {
				model = &m
			}
			if u, ok := mv["usage"]; ok {
				usg := parseUsage(u)
				usage = &usg
			}
			requestID = e.RequestID
		case "system":
			isMeta = boolOr(e.IsMeta)
			if e.Subtype != nil {
				subtype = e.Subtype
				eventData = buildSystemEventData(e)
			}
		}
	}

	if e.EntryType == "user" && e.IsCompactSummary != nil && *e.IsCompactSummary {
		isCompactSummary = true
	}

	if e.EntryType == "progress" {
		s := "progress"
		subtype = &s
		if len(e.Data) > 0 {
			var dm map[string]json.RawMessage
			if json.Unmarshal(e.Data, &dm) == nil {
				if msg, ok := decodeString(dm["message"]); ok {
					content = textContent(msg)
				}
			}
		}
	}

	if e.EntryType == "queue-operation" {
		s := "queue_operation"
		subtype = &s
		eventData = &domain.SystemEventData{
			Subtype:       "queue_operation",
			Operation:     e.Operation,
			QueuedContent: e.Content,
		}
	}

	timestamp := time.Now().UTC().Format(time.RFC3339)
	if e.Timestamp != nil {
		timestamp = *e.Timestamp
	}

	var parentUUID *string
	if isConversational {
		parentUUID = e.ParentUUID
	}

	var compactPtr *bool
	if isCompactSummary {
		t := true
		compactPtr = &t
	}

	return &domain.ParsedMessage{
		UUID:                    uuid,
		ParentUUID:              parentUUID,
		MessageType:             e.EntryType,
		Timestamp:               timestamp,
		Role:                    role,
		Content:                 content,
		Usage:                   usage,
		Model:                   model,
		Cwd:                     e.Cwd,
		GitBranch:               e.GitBranch,
		AgentID:                 e.AgentID,
		IsSidechain:             e.IsSidechain,
		IsMeta:                  isMeta,
		UserType:                e.UserType,
		ToolCalls:               extractToolCalls(content),
		ToolResults:             extractToolResults(content),
		SourceToolUseID:         e.SourceToolUseID,
		SourceToolAssistantUUID: e.SourceToolAssistantUUID,
		ToolUseResult:           e.ToolUseResult,
		IsCompactSummary:        compactPtr,
		RequestID:               requestID,
		Subtype:                 subtype,
		EventData:               eventData,
	}
}

func boolOr(p *bool) bool {
	return p != nil && *p
}
